File: litellm/llms/openai_like/json_loader.py
"""
JSON-based provider configuration loader for OpenAI-compatible providers.
"""

import json
from pathlib import Path
from typing import Final

from litellm._logging import verbose_logger


class SimpleProviderConfig:
    """Simple data class for JSON provider config"""

    def __init__(self, slug: str, data: dict):
        self.slug = slug
        self.base_url = data["base_url"]
        self.api_key_env = data["api_key_env"]
        self.api_base_env = data.get("api_base_env")
        self.base_class = data.get("base_class", "openai_gpt")
        self.param_mappings = data.get("param_mappings", {})
        self.constraints = data.get("constraints", {})
        self.special_handling = data.get("special_handling", {})
        self.supported_endpoints = data.get("supported_endpoints", [])


class JSONProviderRegistry:
    """Load providers from JSON once on import"""

    _providers: dict[str, SimpleProviderConfig] = {}
    _loaded = False

    @classmethod
    def load(cls):
        """Load providers from JSON configuration file"""
        if cls._loaded:
            return

        json_path: Final = Path(__file__).parent / "providers.json"

        if not json_path.exists():
            # No JSON file yet, that's okay
            cls._loaded = True
            return

        try:
            with open(json_path) as f:
                data: Final = json.load(f)

            for slug, config in data.items():
                cls._providers[slug] = SimpleProviderConfig(slug, config)

            cls._loaded = True
        except Exception as e:
            verbose_logger.warning("Warning: Failed to load JSON provider configs: %s", e)
            cls._loaded = True

    @classmethod
    def get(cls, slug: str) -> SimpleProviderConfig | None:
        """Get a provider configuration by slug"""
        return cls._providers.get(slug)

    @classmethod
    def exists(cls, slug: str) -> bool:
        """Check if a provider is defined via JSON"""
        return slug in cls._providers

    @classmethod
    def get_by_base_url(cls, base_url: str) -> SimpleProviderConfig | None:
        """Get a provider configuration by its default base url"""
        return next((provider for provider in cls._providers.values() if provider.base_url == base_url), None)

    @classmethod
    def supports_responses_api(cls, slug: str) -> bool:
        """Check if a JSON provider supports the Responses API"""
        provider: Final = cls._providers.get(slug)
        if provider is None:
            return False
        return "/v1/responses" in provider.supported_endpoints

    @classmethod
    def list_providers(cls) -> list:
        """List all registered provider slugs"""
        return list(cls._providers.keys())


# Load on import
JSONProviderRegistry.load()

File: litellm-rust/crates/llms/src/openai_like/json_loader.rs
use std::collections::BTreeMap;

use crate::base_llm::translation::TranslationError;

pub enum ProviderBaseClass {
    OpenAiGpt,
    OpenAiLikeChat,
}

pub struct ProviderConstraints {
    pub temperature_min: Option<f64>,
    pub temperature_max: Option<f64>,
    pub temperature_min_with_n_gt_1: Option<f64>,
}

pub struct ProviderSpecialHandling {
    pub convert_content_list_to_string: bool,
    pub force_store_false: bool,
}

pub struct SimpleProviderConfig {
    pub slug: String,
    pub base_url: Option<String>,
    pub api_key_env: String,
    pub api_base_env: Option<String>,
    pub base_class: ProviderBaseClass,
    pub param_mappings: BTreeMap<String, String>,
    pub constraints: ProviderConstraints,
    pub special_handling: ProviderSpecialHandling,
    pub supported_endpoints: Box<[String]>,
}

pub struct JsonProviderRegistry {
    _providers: BTreeMap<String, SimpleProviderConfig>,
}

impl JsonProviderRegistry {
    pub fn load(_definitions: &[u8]) -> Result<Self, TranslationError> {
        todo!()
    }

    pub fn get(&self, _slug: &str) -> Option<&SimpleProviderConfig> {
        todo!()
    }

    pub fn exists(&self, _slug: &str) -> bool {
        todo!()
    }

    pub fn get_by_base_url(&self, _base_url: &str) -> Option<&SimpleProviderConfig> {
        todo!()
    }

    pub fn supports_responses_api(&self, _slug: &str) -> bool {
        todo!()
    }

    pub fn list_providers(&self) -> Box<[&str]> {
        todo!()
    }
}
